### PR TITLE
Static Web Publishing with IPFS - Simpler and Decentralized

### PR DESCRIPTION
**Pull request by diasdavid**

### CODE REVIEW

File: src/10-ipscend-update/index.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+** The case for this tool 
+
+# Using ipscend
+
+- [ ] asciinema / video
+
+# Users friendly
+
+- [ ] explain naming problem
+- [ ] explain how IPFS public gateways can be the proxies for users seemlessly (with graphs!)
```
Comment (lgierth) [inline src/10-ipscend-update/index.md]: we could reference the firefox addon here

File: src/10-ipscend-update/index.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+** The case for this tool 
+
+# Using ipscend
+
+- [ ] asciinema / video
+
+# Users friendly
+
+- [ ] explain naming problem
```
Comment (lgierth) [inline src/10-ipscend-update/index.md]: naming problem, is this where the user gets a hash and puts it into a dnslink or IPNS?

File: src/10-ipscend-update/index.md (hunk 1 above)
Comment (diasdavid) [inline src/10-ipscend-update/index.md]: Doesn't it have a different purpose? Here would be explaining how we can get DNS names to load IPFS hashes through using IPFS public gateways as a proxy

File: src/10-ipscend-update/index.md (hunk 2 above)
Comment (diasdavid) [inline src/10-ipscend-update/index.md]: exactly :)

**Comment by RichardLitt** [comment]
Ping me before merging. I'm going to comb over this again when ya'll are done.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+---
+
+> - Don't you hate it when your links break?
+> - Don't you hate it when your webhost gets attacked and your websites go down?
+> - Don't you hate managing servers just to self-host a webpage?
+> - Don't you hate it when you switch cloud providers and all the data links change?
+> - Don't you hate it when people relying on your content get broken embeds?
+
+> NOTE: I think it'd be better to start stronger than rhetorical questions. For instance:    
+> The web is broken. Links break; webhosts get attacked, and websites go down (often permanently). As a web developer, you have to by a sysadmin just to host a webpage. If you switch cloud providers, links break. Even with a fully functional website, more issues might pop up; link rot, or broken embeds for third parties serving your content. It's 2016, and not all is as it should be.
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: We're going to have to go with one of these, or the other. The questions, or the paragraph.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+"I just wanted to publish my personal website!" It is not a mystery why  most people use Facebook, Twitter, Medium, and Tumblr to publish and host their public personas, their blogs, and all their thoughts.
+
+Suggestion: This is where IPFS, the permanent web, comes in. We've been working on a solution to these issues, imagining a word where it is easier to publish web content; where it easier to host and maintain it; where links don't break or change; where old versions didn't disappear in moment; and where all of this is as simple as tweeting. All at once.
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: Same with this paragraph. I think it is stronger than the questions.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+> - What if it was way easier to publish web content?
+> - What if it was way easier to host and maintain it?
+> - What if by design, links didn't break?
+> - What if by design, links didn't change when moving hosts?
+> - What if by design, old versions were all still accessible?
+> - What if managing all this was as simple as tweeting?
+
+(NOTE: Again, I think these questions come off as a bit weak. People skim through repetitive questions, only reading the right edge and not getting the rhetorical effect that you'd get in speech. Rephrase, like above.)
+
+We don't have all the solutions working yet, but we're working hard to get there. Comparatively, the IPFS Project is very young and we're still doing out all the groundwork and creating the new plumbing infrastructure. However, we can already make a big dent into these annoying problems, at least for developers and users familiar with:
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: First time we mention the IPFS project. Obviously, this is on our website, but just saying. Link, maybe.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+collection: posts
+---
+
+> - Don't you hate it when your links break?
+> - Don't you hate it when your web host gets attacked and your websites go down?
+> - Don't you hate managing servers just to self-host a web page?
+> - Don't you hate it when you switch cloud providers and all the data links change?
+> - Don't you hate it when people relying on your content get broken embeds?
+
+> NOTE: I think it'd be better to start stronger than rhetorical questions. For instance:
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: We're going to have to go with one of these, or the other. The questions, or the paragraph.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+"I just wanted to publish my personal website!" It is not a mystery why most people use Facebook, Twitter, Medium, and Tumblr to publish and host their public personas, their blogs, and all their thoughts.
+
+Suggestion: This is where IPFS, the permanent web, comes in. We've been working on a solution to these issues, imagining a word where it is easier to publish web content; where it easier to host and maintain it; where links don't break or change; where old versions didn't disappear in moment; and where all of this is as simple as tweeting. All at once.
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: Same with this paragraph. I think it is stronger than the questions.

Again, I think these questions come off as a bit weak. People skim through repetitive questions, only reading the right edge and not getting the rhetorical effect that you'd get in speech. Rephrase, like above.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+
+#### Timeline view
+
+One other feature that will significantly increase the developer experience with `ipscend` is [**versioning**](https://github.com/ipfs/notes/issues/23). Apps, as with any software, have several iterations across their lifetime. These iterations pack different things, like new features, complete application revamps or sometimes, regressions.
+
+Version Control Systems have enabled developers to work collaboratively on the same code, without wasting time dealing with merge conflicts. We can improve these tools by letting developers, designers and other individuals have more power. Ultimately, they should be able to have quick access to the timeline for their application, to make annotations on the current iterations before a release, to cherry pick which version to be released (using visual rollbacks), to analyze if there have been any regressions in a specific browser, among other things.
+
+Currently we have a `timeline` feature that lets you browse through screenshots of all of the published versions of your application.
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: Needs to be fleshed out.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+
+There is much more to do, more problems to simplify, and lots to implement. We will end this post highlighting a some things we're working on to improve the web publishing experience even further.
+
+#### Timeline view
+
+One other feature that will significantly increase the developer experience with `ipscend` is [**versioning**](https://github.com/ipfs/notes/issues/23). Apps, as with any software, have several iterations across their lifetime. These iterations pack different things, like new features, complete application revamps or sometimes, regressions.
+
+Version Control Systems have enabled developers to work collaboratively on the same code, without wasting time dealing with merge conflicts. We can improve these tools by letting developers, designers and other individuals have more power. Ultimately, they should be able to have quick access to the timeline for their application, to make annotations on the current iterations before a release, to cherry pick which version to be released (using visual rollbacks), to analyze if there have been any regressions in a specific browser, among other things.
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: This paragraph is really information dense. Break it up for each thing web developers could do, and talk about how our efforts there are going.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+
+Currently we have a `timeline` feature that lets you browse through screenshots of all of the published versions of your application.
+
+![](http://zippy.gfycat.com/TameDampKob.gif)
+
+#### Extending current VCS with IPFS and IPLD
+
+The [IPLD (InterPlanetary Linked Data)](https://github.com/ipfs/specs/blob/master/merkledag/ipld.md) data model will enable current Version Control Systems to be extended so that their data structures can live inside the IPFS network, without having to change how that they work. What this means is that we will be able to have one single source tree, where releases will be one of the iterations and where users will be able to leverage the last .....
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: To Do: Finish

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+
+You'll find more information about the dnslink mechanism in the [go-dnslink README file](https://github.com/jbenet/go-dnslink). We'll cover it in a more in-depth blog post very soon. We're  also working hard to add support for other name systems  -- for example, Namecoin -- too.
+
+#### Step 5: Backing up your content elsewhere.
+
+One of the most important things to understand about IPFS is how content moves around the network. Like many peer-to-peer and distributed systems, peers can access and download content directly from each other. But unlike many peer-to-peer systems, IPFS does not automatically push your content to others to back it up. This is so that you can use IPFS to store, address, and move your own data **without** requiring you to store other people's data. This is an important distinction! We sacrifice built-in replication for a richer model that matches how people feel about content. Most individuals and businesses _would never_ use a protocol if it _forced_ them to store other people's unknown data. Because who knows, it might be something illegal; protocols that force that are unlikely to be used massively.
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: Not sure any of this bold and italics are necessary.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+
+**So then how do we replicate, disperse, and backup data?** The short answer is: in a protocol layered _on top of_ IPFS. IPFS tells you how to access and transfer data, and other protocols layer on top of IPFS to provide replication and redundancy guarantees. This closely matches how the world's individuals, organizations, and services administer and replicate their data today. HTTP does not guarantee your data will be around, it just tells you how to access and transfer it. Users and HTTP websites rely on shared systems, or services like Amazon S3, to back up their data. For IPFS, there are emerging protocols and applications to organize data replication. Some of them, like pincoop and ... TODO Finish
```
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: Needs to be finished.

**Comment by RichardLitt** [comment]
The piece is a bit loose at the beginning, very casual during the steps, and then really dense at the end. Would be good to tighten up the beginning a bit so that users don't get jarred towards the end. Also, add some sign posting at the beginning, saying what we are going to do and what we will talk about at the end, too. Otherwise, people will just think it is a tutorial, and not also an exposé of future plans.

File: src/10-webpage-publishing-with-ipfs/index.md [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+---
+
+> - Don't you hate it when your links break?
+> - Don't you hate it when your web host gets attacked and your websites go down?
+> - Don't you hate managing servers just to self-host a web page?
+> - Don't you hate it when you switch cloud providers and all the data links change?
+> - Don't you hate it when people relying on your content get broken embeds?
+
+> NOTE: I think it'd be better to start stronger than rhetorical questions. For instance:    
+> The web is broken. Links break; web hosts get attacked, and websites go down (often permanently). As a web developer, you have to by a sysadmin just to host a web page. If you switch cloud providers, links break. Even with a fully functional website, more issues might pop up; link rot, or broken embeds for third parties serving your content. It's 2016, and not all is as it should be.
```
Comment (jbenet) [inline src/10-webpage-publishing-with-ipfs/index.md]: "have to be a sysadmin"

Comment (jbenet) [inline src/10-webpage-publishing-with-ipfs/index.md]: "If you switch cloud providers, links often break." (add "often")

File: src/10-webpage-publishing-with-ipfs/index.md (hunk 6 above)
Comment (jbenet) [inline src/10-webpage-publishing-with-ipfs/index.md]: Yeah agreed. not sure which one is stronger. others' opinions?

File: src/10-webpage-publishing-with-ipfs/index.md (hunk 13 above)
Comment (RichardLitt) [inline src/10-webpage-publishing-with-ipfs/index.md]: Good catch. Should we keep this paragraph?

File: src/10-webpage-publishing-with-ipfs/index.md (hunk 6 above)
Comment (diasdavid) [inline src/10-webpage-publishing-with-ipfs/index.md]: I like the direct pulls the better. Also, saying that the "The web is broken" is kind of an hard statement and a big hit on everyone that puts so much work on it

**Comment by diasdavid** [comment]
- [ ] (from @jbenet ) we need a part for ​_updating_​ the site. which works, given we use dnslink. we can do it in a "next steps" section. and reference ipns

**Comment by jbenet** [comment]
I updated the post. It still needs some images. And i need to improve the top image. (make it from teh proper stock photo).

**Comment by jbenet** [comment]
From here out, let's just stick with this, no major wording changes. we still need to:

- [ ] do one more pass
  - [ ] correct typos
  - [ ] correct grammar
  - [ ] make sure all the commands work exactly right.
- [ ] make proper topbar image
- [ ] add other images:
  - [ ] an image for links broken (intro)
  - [ ] an image for DNS naming
  - [ ] an image for pinning
  - [ ] an image for successful web publishing
- [ ] download all images and host them in this repo

@RichardLitt if you could help with some, A+.

File: src/10-static-web-publishing-with-ipfs/index.md [hunk 14]
```diff
@@ -0,0 +1,537 @@
...
+collection: posts
+---
+
+# Static Web Publishing with IPFS - Simpler and Decentralized
+
+![](https://www.evernote.com/l/AMYNVaoyi89PQqQU2myImhtVuG628DRn6zYB/image.png)
+
+> "I've always believed that everyone involved in "publishing" should know how to set up a server." -- Joi Ito ([source](http://joi.ito.com/weblog/2016/02/21/a-challenge-fro.html), [archive](https://web.archive.org/web/20160223200231/http://joi.ito.com/weblog/2016/02/21/a-challenge-fro.html), [ipfs](<ipfs-link>))
+>
+> Setting up a server is hard. You have to be a sysadmin just to host a single web page. Even with a fully functional website, more issues pop up: [websites go down](https://blogs.loc.gov/digitalpreservation/2011/11/the-average-lifespan-of-a-webpage/), [hosts get attacked](http://arstechnica.com/security/2015/03/github-battles-largest-ddos-in-sites-history-targeted-at-anti-censorship-tools/), [services shut down](http://arstechnica.com/business/2009/04/geocities-to-close-after-15-years-of-aesthetic-awesomeness/), [links rot](https://en.wikipedia.org/wiki/Link_rot). It's 2016, and publishing even a static web page is way too hard.
```
Comment (diasdavid) [inline src/10-static-web-publishing-with-ipfs/index.md]: needs to have the `>` removed

File: src/10-static-web-publishing-with-ipfs/index.md [hunk 15]
```diff
@@ -0,0 +1,537 @@
...
+
+
+Publishing web content has become very complex, confusing, and [precarious](http://journalistsresource.org/studies/society/internet/website-linking-best-practices-media-online-publishers). Even publishing _static_ web content -- simple webpages, images, videos, or documents -- is way harder than it should be. The "easy way" means depending on services that may change things under you or shut down completely. Hundreds of millions of people have been served, sometimes overnight, with a "Sorry, we're going in a different direction" message. The lucky ones get a dump of their data and a short timeline to "get their affairs in order" and rehost their content. This means hard work rebuilding the site elsewhere, and -- for those who were not [extremely careful with domain names and URLs up-front](https://www.w3.org/Provider/Style/URI) -- attempting to update soon-to-be-broken links all over the web. Like changing mailing addresses after moving, most people resign themselves to some things breaking forever.
```
Comment (diasdavid) [inline src/10-static-web-publishing-with-ipfs/index.md]: :+1: I like this paragraph a lot

File: src/10-static-web-publishing-with-ipfs/index.md [hunk 16]
```diff
@@ -0,0 +1,537 @@
...
+> - What if you could count on your links _decades_ from now, with various versions still accessible?
+> - What if "hosting" did not also mean "dependence on disappearing services"?
+> - What if cloud infrastructure worked more like the power grid, where switching providers is transparent?
+> - What if _you_ controlled?
+> - What if names (like DNS domains) mapped straight to content, not servers?
+> - What if we didn't _have to_ rely on centralized naming (DNS)?
+> - What if webapps and websites worked offline, or in local area networks?
+> - What if -- the kicker -- managing all this was as simple as tweeting?
+
+These are all hard problems, and full solutions will take time to build. We don't have everything working yet, IPFS is very young -- we're still implementing fundamental infrastructure and laying the groundwork for user-facing applications. But we have already put dents into many annoying problems, and this post shows off some of the tooling we are building.
```
Comment (diasdavid) [inline src/10-static-web-publishing-with-ipfs/index.md]: I would remove the `still`, makes it sounds that we are taking more time than promised/expected (which is not actually true)

File: src/10-static-web-publishing-with-ipfs/index.md [hunk 17]
```diff
@@ -0,0 +1,537 @@
...
+We will work on simplifying these processes even further, but for now, we hope that this is enough to help many people to re-host their web content in easier, more resilient, and permanent ways.
+
+
+**Tutorial Steps**:
+- [Step 0: Installing `ipfs` and running a node](#step-0-installing-ipfs-and-running-a-node)
+- [Step 1: Prepare the static website to publish](#step-1-prepare-the-static-website-to-publish)
+- [Step 2: ipfs add to publish a website at an IPFS address](#step-2-ipfs-add-to-publish-a-website-at-an-ipfs-address)
+- [Step 3 (optional): ipscend - tooling to publish your page and track versions](#step-3-optional-ipscend-tooling-to-publish-your-page-and-track-versions)
+- [Step 4 (optional): human readable naming with DNS and `dnslink`, and familiar HTTP links](#step-4-optional-human-readable-naming-with-dns-and-dnslink-and-familiar-http-links)
+- [Step 5: Backing up your content elsewhere (pinning)](step-5-backing-up-your-content-elsewhere-pinning)
```
Comment (diasdavid) [inline src/10-static-web-publishing-with-ipfs/index.md]: also optional, no?

File: src/10-static-web-publishing-with-ipfs/index.md [hunk 18]
```diff
@@ -0,0 +1,537 @@
...
+```
+
+If `ipfs get` fails, you'll need to make sure your ipfs node is running, with `ipfs daemon`, or see **Step 0** above.
+
+
+**(Optional)** If you want to preview this website, you can try any HTTP server. For example, the `node-static` module (available on npm):
+
+```sh
+> npm install --global node-static
+> cd static-webpage-example
```
Comment (diasdavid) [inline src/10-static-web-publishing-with-ipfs/index.md]: no need for this `cd` anymore

File: src/10-static-web-publishing-with-ipfs/index.md [hunk 19]
```diff
@@ -0,0 +1,537 @@
...
+> static src
```
Comment (diasdavid) [inline src/10-static-web-publishing-with-ipfs/index.md]: `static website`

File: src/10-static-web-publishing-with-ipfs/index.md [hunk 20]
```diff
@@ -0,0 +1,537 @@
...
+ipfs.io.                120     IN      A       178.62.61.185
+ipfs.io.                120     IN      A       104.236.151.122
+ipfs.io.                120     IN      A       178.62.158.247
+ipfs.io.                120     IN      A       104.236.176.52
+ipfs.io.                120     IN      A       162.243.248.213
+ipfs.io.                120     IN      A       104.236.179.241
+ipfs.io.                120     IN      A       104.236.76.40
+...
+
+> dig +short A dist.ipfs.io
```
Comment (diasdavid) [inline src/10-static-web-publishing-with-ipfs/index.md]: do we really want to show the normal and short version?

File: src/10-static-web-publishing-with-ipfs/index.md [hunk 21]
```diff
@@ -0,0 +1,537 @@
...
+- (federations) a few different tools are emerging to bind together sets of nodes across different organizations -- i.e. across trust boundaries. Two examples are [ipfs-persistence-consortium](https://github.com/pipermerriam/ipfs-persistence-consortium), and [pincoop](https://github.com/victorbjelkholm/pincoop). These tools can use testers and contributions.
+- (bots) we are creating a set of "pinbots" -- bots that pin at the request of their users -- for [IRC](https://github.com/ipfs/pinbot-irc), Twitter, GitHub, and more. If you're interested in other channels, please let us know. [Hubot](https://hubot.github.com/) makes it easy to add `pinbots` to various services.
+- (cloud service) [Protocol Labs](http://ipn.io) is creating a cloud service on top of existing cloud providers. Stay tuned for more news on this.
+- (incentivized protocols) [Protocol Labs](http://ipn.io) is also creating [Filecoin](http://filecoin.io), a cryptocurrency incentivized storage network on top of IPFS. This will be the best way to pin long-term. If you're interested in building Filecoin with Protocol Labs, [reach out](http://ipn.io/join).
+
+
```
Comment (diasdavid) [inline src/10-static-web-publishing-with-ipfs/index.md]: 🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟🌟 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

File: src/10-static-web-publishing-with-ipfs/index.md (hunk 18 above)
Comment (jbenet) [inline src/10-static-web-publishing-with-ipfs/index.md]: can you fix + push it?

File: src/10-static-web-publishing-with-ipfs/index.md (hunk 17 above)
Comment (jbenet) [inline src/10-static-web-publishing-with-ipfs/index.md]: eh no, i'd say they should do this

Comment (jbenet) [inline src/10-static-web-publishing-with-ipfs/index.md]: but sure i guess.

File: src/10-static-web-publishing-with-ipfs/index.md (hunk 16 above)
Comment (jbenet) [inline src/10-static-web-publishing-with-ipfs/index.md]: ok sounds good

File: src/10-static-web-publishing-with-ipfs/index.md (hunk 20 above)
Comment (jbenet) [inline src/10-static-web-publishing-with-ipfs/index.md]: ok just the short then

File: src/10-static-web-publishing-with-ipfs/index.md (hunk 18 above)
Comment (diasdavid) [inline src/10-static-web-publishing-with-ipfs/index.md]: I can :)

**Comment by RichardLitt** [comment]
To do:
- [ ] @RichardLitt overview again
- [ ] Pin cat image on both networks
- [ ] @whyrusleeping: "we advise the user to try out node-static to preview their webpage on port 8080. But if they followed the instructions correctly, they have an ipfs node running a gateway there." Fix.

**Comment by RichardLitt** [comment]
I was unable to do the following: `ipfs get -o website/cat.jpg /ipfs/QmZVLbEzkoZBMMzBsLfY4d86NFMXo1yBBhsqucd7w7VnGV/cat.jpg`. No one responded on IRC, so I am stopping going through the tutorial at the moment. No idea why `get` fails. Using the online gateway worked fine.

**Comment by RichardLitt** [comment]
Waiting on designs. Review once designs have been had.

**Comment by Kubuxu** [comment]
`ipscend` needs to be fixed before this can be released.

**Comment by diasdavid** [comment]
_Closing ancient posts that never got posted._

**Pull request closed without merging**